### PR TITLE
[dynamo] propagate tensor metadata on Tensor.__setitem__(tensor)

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161036

Fixes silent incorrectness for autograd function tracing, where we rely on FakeTensor metadata (requires_grad) to determine whether to HOP or not: https://github.com/pytorch/pytorch/blob/5ee464db5c4293ac09521f9069fa7d2106680a7f/torch/_dynamo/variables/misc.py#L671

Stared at this with @anijain2305 yesterday, `Tensor.__setitem__` can update tensor metadata, and we can just run the fake prop and extract the output metadata from the updated FakeTensor.

FIXES https://github.com/pytorch/pytorch/issues/160901

It should also be the root cause behind the issue in https://github.com/pytorch/torchtitan/pull/1604 @bdhirsh  @ruisizhang123

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela @mlazos

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -2060,32 +2060,8 @@ def wrap_tensor(self, value: torch.Tensor):
...
-            # __torch_function__
-            #
-            # To simplify things for now, the __dict__ tracking bits haven't
-            # been implemented yet, but they can be added into this design at
-            # a later point in time.
-            subclass_type = None
-        else:
-            subclass_type = type(value)
+        subclass_type = infer_subclass_type(value)
+        if subclass_type is not None:
```
Comment (bdhirsh) [inline torch/_dynamo/variables/builder.py:2064]: is this change related? (i thought this was just a refactor but the branch here looks new)

Comment (xmfan) [reply torch/_dynamo/variables/builder.py:2064]: its the same logic as before, this else path additionally adds guards

File: test/dynamo/test_repros.py [hunk 2]
```diff
@@ -7140,6 +7140,39 @@ def f(eye, out):
...
+
+        x1, y1 = inputs()
+        fn(x1, y1).sum().backward()
+        self.assertTrue(x1.requires_grad)
+
+        x2, y2 = inputs()
+        torch.compile(fn, backend="eager")(x2, y2).sum().backward()
+        self.assertTrue(x2.requires_grad)
+
+        self.assertEqual(x1.grad, x2.grad)
```
Comment (bdhirsh) [inline test/dynamo/test_repros.py]: I tried running your test locally without your changes, and it looks like this test passes without the changes in this PR - it turns out that `x` is a non-leaf tensor after the mutation (`x.grad_fn` is CopySlices`), so `x.grad` is `None` in both examples.

You probably need to assert on `self.assertEqual(y1.grad, y2.grad)`? since the "wrong gradients" that we are computing today should propagate into y2.grad

File: torch/_dynamo/variables/tensor.py [hunk 3]
```diff
@@ -1090,6 +1090,28 @@ def method___setitem__(self, key, value):
...
+                get_fake_value(proxy.node, tx, allow_non_graph_fake=False)
+
+            example_value = self.proxy.node.meta.get("example_value")
+            from .builder import get_specialized_props, infer_subclass_type
+
+            specialized_props = get_specialized_props(
+                target_cls, tx, example_value, infer_subclass_type(example_value)
+            )
+            for k, v in specialized_props.items():
+                setattr(self, k, v)
```
Comment (bdhirsh) [inline torch/_dynamo/variables/tensor.py:1115]: We might need to audit a few other places in dynamo too. For example, this will also change the requires_gradness of `x`:
```
x = torch.randn(4)
y = torch.randn(4, requires_grad=True)
# x.requires_grad is now True
x.add_(y)
```

so any tensor data mutation ops will need the same treatment

File: test/dynamo/test_repros.py (hunk 2 above)
Comment (xmfan) [reply test/dynamo/test_repros.py]: Thanks for the catch, let me just add the original repro's permutations...

File: torch/_dynamo/variables/tensor.py (hunk 3 above)
Comment (bdhirsh) [reply torch/_dynamo/variables/tensor.py:1115]: Confirmed locally - i tried tweaking your input_fn to use add_ and hit the same correctness issue when asserting on y.grad:
```
        def fn(x, y):
            x.add_(y)
            return MyFn.apply(x)
```

**Review by bdhirsh — COMMENTED** [review]
Left a few comments - nice find!

**Comment by xmfan** [comment]
hf_Bart fails on base commit

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/tensor.py (hunk 3 above)
Comment (bdhirsh) [reply torch/_dynamo/variables/tensor.py:1115]: hmm does this PR also fix the problem for `add_`? it's not clear to me where (if not no worries, but can you file another issue?)
